### PR TITLE
Fix a flaky test for gdd/dist-deadlock-upsert

### PR DESCRIPTION
**Pull request by gfphoenix78**
When to run GDD probe is undermined, but it is import for the test
gdd/dist-deadlock-upsert. If the GDD probe runs immediately after
the 2 inter-dead-locked transactions, one of the transactions will
be killed. The isolation2 framework consider the transaction being
blocked if the transaction doesn't finished in 0.5 second. So, if
the killed transaction is too early to be aborted, the test framework
sees no dead lock.

Analyzed-by: Gang Xiong <gxiong@pivotal.io>

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/test/isolation2/sql/gdd/dist-deadlock-upsert.sql [hunk 1]
```diff
@@ -19,12 +19,20 @@ INSERT INTO t_upsert (id, val) SELECT i, i FROM generate_series(1, 100) i;
...
+select gp_wait_until_triggered_fault('gdd_probe', 1, dbid)
+  from gp_segment_configuration where content=-1 and role='p';
 -- seg 0: con20 ==> con10, xid lock
 20&: INSERT INTO t_upsert VALUES (segid(0,1), segid(0,1)) on conflict (id, val) do update set val = 666;
 
 -- seg 1: con10 ~~> con20, tuple lock
 10&: INSERT INTO t_upsert VALUES (segid(1,1), segid(1,1)) on conflict (id, val) do update set val = 777;
 
+-- isolation2 framework requires at least 0.5 second to infer whether a transaction is blocked.
+select pg_sleep(0.6);
```
Comment (xiong-gang) [inline src/test/isolation2/sql/gdd/dist-deadlock-upsert.sql:33]: we already 'suspend' the prober, do we need to sleep here?

Comment (gfphoenix78) [reply src/test/isolation2/sql/gdd/dist-deadlock-upsert.sql:33]: If we don't sleep some time more than 0.5 second, one of the blocked transaction may be killed in less than 0.5 second when the GDD starts to probe.

Comment (kainwen) [reply src/test/isolation2/sql/gdd/dist-deadlock-upsert.sql:33]: > If we don't sleep some time more than 0.5 second, one of the blocked transaction may be killed in less than 0.5 second when the GDD starts to probe.

The 0.5 s is executed in the handling `&`, not here.

Comment (gfphoenix78) [reply src/test/isolation2/sql/gdd/dist-deadlock-upsert.sql:33]: I know. If the GDD probe runs immediately after the second INSERT started, one of 2 transactions will be killed and it probably runs in less 0.5 second. So the isolation2 framework may see no blocked transactions.

Comment (gfphoenix78) [reply src/test/isolation2/sql/gdd/dist-deadlock-upsert.sql:33]: @xiong-gang @kainwen You're right. I thought the wait process is in background, but it isn't.

Comment (gfphoenix78) [reply src/test/isolation2/sql/gdd/dist-deadlock-upsert.sql]: Removed.

**Comment by xiong-gang** [comment]
+1 thanks for fixing it

**Comment by ashwinstar** [comment]
Again commit message only explained the problem but not the fix. Please lets get in habit of explaining problem and fix both in commit message to make it easy for reviews and for git annotates.

File: src/test/isolation2/isolation2_schedule [hunk 2]
```diff
@@ -25,7 +25,8 @@ test: packcore
 # Tests on global deadlock detector
 test: gdd/prepare
 test: gdd/concurrent_update
-test: gdd/dist-deadlock-01 gdd/dist-deadlock-04 gdd/dist-deadlock-05 gdd/dist-deadlock-06 gdd/dist-deadlock-07 gdd/dist-deadlock-102 gdd/dist-deadlock-103 gdd/dist-deadlock-104 gdd/dist-deadlock-106 gdd/dist-deadlock-upsert gdd/non-lock-105
+test: gdd/dist-deadlock-01 gdd/dist-deadlock-04 gdd/dist-deadlock-05 gdd/dist-deadlock-06 gdd/dist-deadlock-07 gdd/dist-deadlock-102 gdd/dist-deadlock-103 gdd/dist-deadlock-104 gdd/dist-deadlock-106 gdd/non-lock-105
+test: gdd/dist-deadlock-upsert
```
Comment (ashwinstar) [inline src/test/isolation2/isolation2_schedule:29]: Would have been good to comment why this can't be run in parallel.

**Comment by gfphoenix78** [comment]
> Again commit message only explained the problem but not the fix. Please lets get in habit of explaining problem and fix both in commit message to make it easy for reviews and for git annotates.

Thank you, Ashwin. It's a good idea. I'll follow the habit in the later commits.

File: src/test/isolation2/isolation2_schedule (hunk 2 above)
Comment (gfphoenix78) [reply src/test/isolation2/isolation2_schedule:29]: Yes, I should put the reason in the schedule file.
It uses fault injector. The check script requires it to run not in parallel. If someone put the test in parallel, he/she will get an error before running the ICW test.

**Pull request merged**